src/gsy_e/models/strategy/energy_parameters/heatpump/cop_models/cop_models.py

```diff
@@ -0,0 +1,141 @@
+    COPModelType.ELCO_AEROTOP_G07_14M: "Elco_Aerotop_G07-14M_model_parameters.json",
+    COPModelType.HOVAL_ULTRASOURCE_B_COMFORT_C11: "hoval_UltraSource_B_comfort_C_11_model_"
+    "parameters.json",
+}
+
+
+class BaseCOPModel:
+    """Base clas for COP models"""
+
+    @abstractmethod
+    def calc_cop(self, source_temp_C: float, tank_temp_C: float, heat_demand_kW: Optional[float]):
+        """Return COP value for provided inputs"""
+
+
+class IndividualCOPModel(BaseCOPModel):
+    """Handles cop models for specific heat pump models"""
+
+    def __init__(self, model_type: COPModelType):
+        with open(
+            os.path.join(MODEL_FILE_DIR, MODEL_TYPE_FILENAME_MAPPING[model_type]),
+            "r",
+            encoding="utf-8",
+        ) as fp:
+            self._model = json.load(fp)
+        self.model_type = model_type
+
+    def _calc_power(self, source_temp_C: float, tank_temp_C: float, heat_demand_kW: float):
+        CAPFT = (
+            self._model["CAPFT"][0]
+            + self._model["CAPFT"][1] * source_temp_C
+            + self._model["CAPFT"][3] * source_temp_C**2
+            + self._model["CAPFT"][2] * tank_temp_C
+            + self._model["CAPFT"][5] * tank_temp_C**2
+            + self._model["CAPFT"][4] * source_temp_C * tank_temp_C
+        )
+
+        HEIRFT = (
+            self._model["HEIRFT"][0]
+            + self._model["HEIRFT"][1] * source_temp_C
+            + self._model["HEIRFT"][3] * source_temp_C**2
+            + self._model["HEIRFT"][2] * tank_temp_C
+            + self._model["HEIRFT"][5] * tank_temp_C**2
+            + self._model["HEIRFT"][4] * source_temp_C * tank_temp_C
+        )
+
+        # Partial Load Ratio (PLR)
+        PLR = heat_demand_kW / (self._model["Qref"] * CAPFT)
+
+        # HEIRFPLR calculation
+        HEIRFPLR = (
+            self._model["HEIRFPLR"][0]
+            + self._model["HEIRFPLR"][1] * PLR
+            + self._model["HEIRFPLR"][2] * PLR**2
+        )
+
+        # Power consumption (P) calculation
+        return self._model["Pref"] * CAPFT * HEIRFT * HEIRFPLR
+
+    def calc_cop(self, source_temp_C: float, tank_temp_C: float, heat_demand_kW: float):
+        assert heat_demand_kW is not None, "heat demand should be provided"
+        if heat_demand_kW == 0:
+            return 0
+        electrical_power_kW = self._calc_power(source_temp_C, tank_temp_C, heat_demand_kW)
+        if electrical_power_kW <= 0:
+            log.debug(
+                "calculated power is negative: "
+                "hp model: %s  source_temp: %s, "
+                "tank_temp: %s, heat_demand_kW: %s, calculated power: %s",
+                self.model_type.name,
+                round(source_temp_C, 2),
+                round(tank_temp_C, 2),
+                round(heat_demand_kW, 2),
+                round(electrical_power_kW, 2),
+            )
+            return 0
+        cop = heat_demand_kW / electrical_power_kW
+        if cop > 6:
+            log.debug(
```

Comment:
Same here, an error log and a reference to the configuration uuid will be nice to have